File: 剑指12.java
class Solution {
    public boolean exist(char[][] board, String word) {
        char[] words = word.toCharArray();
        for(int i = 0; i < board.length; i++){
            for (int j = 0; j < board[0].length; j++){
                if (dfs(board, words, i, j, 0)) return true;
            }
        }
        return false;
    }
    boolean dfs(char[][] board, char[] words, int i, int j, int k){
        if(i<0 || i > board.length-1 || j < 0 || j > board[0].length-1 || board[i][j] != words[k]){
            return false;
        }
        if( k == words.length-1) return true;
        board[i][j] = '\0';
        boolean result = dfs(board, words, i+1, j, k+1) || dfs(board, words, i-1, j, k+1) ||
                         dfs(board, words, i, j+1, k+1) || dfs(board, words, i, j-1, k+1);
        board[i][j] = words[k];
        return result;
    }
}

File: 剑指12.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        def dfs(i, j, cur):
            # 先判断不成立的
            if not 0 <= i < len(board) or not 0 <= j < len(board[0]) or not board[i][j] == word[cur]:
                return False
            # 成立的
            if cur == len(word)-1:return True
            temp = board[i][j]  # 保存目前遍历的字母
            board[i][j] = ''    # 去除
            result = dfs(i+1, j, cur+1) or dfs(i, j + 1, cur+1) or dfs(i-1,j,cur+1) or dfs(i, j-1,cur+1)
            board[i][j] = temp  # 恢复
            return result
        for i in range(len(board)):
            for j in range(len(board[0])):
                if dfs(i, j, 0): return True
        return False
